DasharoPayloadPkg/Library/FmpDeviceSmmLib/FmpDeviceSmmLib.c

```
@@ -956,9 +958,13 @@ FmpDeviceSetImageWithStatus (
     return Status;
   }
 
-  ReadBuffer = AllocatePool (BlockSize);
-  if (ReadBuffer == NULL) {
-    DEBUG ((DEBUG_ERROR, "%a(): failed to allocate read buffer\n", __FUNCTION__));
+  CurrentImage = ReadCurrentFirmware ();

```

How long does the reading process take? If it is noticeable maybe the progress bar should be moved by few percents (before and after reading) to calm the users.

---

About 10 seconds.  Took a while to figure this out because I assumed things to work on MSI, but this required trying it out there and nothing worked.  Previously, the reads could fail silently and they actually did.

Turned out that first 4 MiB of flash aren't mmapped, only "bios" region is.  Not sure how to resolve this properly, locally I've added `4 * MiB` to size at https://github.com/Dasharo/coreboot/blob/bd294ab1d22b4010fafa0120e2eda26964316096/src/soc/intel/common/block/fast_spi/mmap_boot.c#L123 which made the whole flash available to coreboot.  I guess this only affects reads as erases or writes didn't fail.

Made this read to move progress by about 5%: https://github.com/Dasharo/edk2/compare/ab54493809413c1639ff172e219168385a576c97..a53ef1c4ebc4be7222fa234b1ce6cb61e6e4a3ee#diff-b33849bc60c6d8a499fabcaf6d579fa92740f3e0e5b0ebb22c24db8af56d307f

---

> Made this read to move progress by about 5%

That part is resolved.

> Turned out that first 4 MiB of flash aren't mmapped, only "bios" region is. Not sure how to resolve this properly, locally I've added 4 * MiB to size at https://github.com/Dasharo/coreboot/blob/bd294ab1d22b4010fafa0120e2eda26964316096/src/soc/intel/common/block/fast_spi/mmap_boot.c#L123 which made the whole flash available to coreboot.

Interesting, especially since it then creates a device for whole `CONFIG_ROM_SIZE` in https://github.com/Dasharo/coreboot/blame/bd294ab1d22b4010fafa0120e2eda26964316096/src/soc/intel/common/block/fast_spi/mmap_boot.c#L134. There is no discussion as to size on gerrit, so perhaps this mismatch wasn't deliberate, or maybe it was done so that SPI addresses are correct.

I think the best option would be to add another Kconfig for mmaping whole ROM instead of just BIOS region, and make it selected when capsules are enabled.

---

Anyway, this is coreboot issue, it doesn't block merging this PR.